### DISCUSSION TITLE
Virtual columns, byte indexing, and the semantics of vim.api.nvim_buf_add_highlight

### DISCUSSION THREAD

**Discussion by willruggiano** [neovim/neovim#26143] [General]
I've been using [lir.nvim] for a long time. Recently, after updating my plugins (along with neovim... bad idea), I noticed when I opened a lir buffer that the filetype icons prefixing each line were no longer highlighted. I had never taken the time to familiarize myself with neovim's byte-indexing semantics but on this occasion I had a reason to, because lir.nvim uses `vim.api.nvim_buf_add_highlight` to highlight filetype icons (see screenshots on the repo's homepage). Here is the [relevant code from lir.nvim](https://github.com/tamago324/lir.nvim/blob/master/lua/lir/devicons.lua);

```lua
function devicons.update_highlight(files)
  local col_start, col_end = #" ", ICON_WIDTH + #" "

  a.nvim_buf_clear_namespace(0, ns, 0, -1)
  for i, file in ipairs(files) do
    a.nvim_buf_add_highlight(0, ns, file.devicons.highlight_name, i - 1, col_start, col_end)
  end
end
```

At runtime, the call to `nvim_buf_add_highlight` looks something like:
```
nvim_buf_add_highlight(0, 35, "DevIconLua", 12, 1, 4)
```

Regardless of the icon (i.e. filetype), the `1, 4` (for column start and column end) are always the same. Naively, I changed the dynamic col_start to be a hardcoded 0... problem solved! I thought: maybe the plugin changed and this got messed up? This doesn't seem to be the case. I wrote a little test script to validate what I _thought_ was the correct behavior;

```lua
---- We expect this script to highlight the last three `-` at the start of this line using the ErrorMsg highlight group.

local line = 0 -- "first line"; zero-indexed
local col_start = 2 -- byte indexed, via `col('.')` with cursor on the second `-`
local col_end   = 4 -- byte indexed, via `col('.')` with cursor on the fourth `-`
local bufnr = 0 -- current buffer
local ns = 0 -- create one for me

vim.api.nvim_buf_add_highlight(bufnr, ns, "ErrorMsg", line, col_start, col_end)
```

(`nvim --clean -c source test.lua`)

As the script suggests, I expect the last three `-`s at the start of the first line to get highlighted. But when running this script, the semantics of vim.api.nvim_buf_add_highlight are effectively; `(start, end]`; only the last **two** `-`s get highlighted.

So... what is the correct behavior of `vim.api.nvim_buf_add_highlight` in regards to clusivity of col_start? It appears to be exclusive.

[lir.nvim]: https://github.com/tamago324/lir.nvim

**Comment by zeertzjq** [comment]
Most API functions use 0-based columns, so it's `[start, end)`.

**Reply by willruggiano** [reply]
I guess that's my confusion. This specific API specifies byte indexing, as does (the return value of) |col()|. At least the help txt says this.

With my cursor on the first column, `col('.')` returns 1. Passing 1 as col_start to nvim_buf_add_highlight does not highlight the first column though.

**Reply by zeertzjq** [reply]
You need to subtract 1 from the column returned by `col()`, as most Vimscript builtin functions use 1-based column indexing.
